src/snapred/backend/dao/ingredients/GroceryListItem.py

```diff
@@ -1,37 +1,149 @@
+from snapred.backend.dao.RunConfig import RunConfig
 from snapred.meta.Config import Config
 
 
 class GroceryListItem(BaseModel):
-    """Holds necessary information for a single item in grocery list"""
+    """
+    Holds necessary information for a single item in grocery list
+    """
 
     workspaceType: Literal["nexus", "grouping"]
-    runConfig: Optional[RunConfig]
+    useLiteMode: bool  # indicates if data should be reduced to lite mode
+    # optional loader:
+    # -- "" tells FetchGroceries to choose the loader
     loader: Literal["", "LoadGroupingDefinition", "LoadNexus", "LoadEventNexus", "LoadNexusProcessed"] = ""
-    isLite: Optional[bool]
+    # one of the below must be set -- nexus requires a runNumber, grouping a groupingScheme
+    runNumber: Optional[str]
     groupingScheme: Optional[str]
+    # grouping workspaces require an instrument definition
+    # these indicate which property defines the sourt, and then that source
     instrumentPropertySource: Optional[Literal["InstrumentName", "InstrumentFilename", "InstrumentDonor"]]
     instrumentSource: Optional[str]
+    # if set to False, nexus data will not be loaded in a clean, cached way
+    # this is faster and uses less memory, if you know you only need one copy
     keepItClean: bool = True
 
+    # IPTS is implemented lazily and is not a true model field
+    # the _ipts private field stores its value, and methods handle getting/setting
+    # this can be avoided with @property getter/setter, only supported in pydantic >v2
+    _ipts: str = PrivateAttr()
+
+    def __init__(self, **kwargs):
+        # set the model as usual
+        super().__init__(**kwargs)
+        # also if IPTS was set in construction, set its value
+        object.__setattr__(self, "_ipts", kwargs.get("IPTS"))
+
+    def dict(self, **kwargs) -> Dict[str, Any]:  # noqa: A003, ARG002
+        # return the usual dictionary
+        ans = super().dict(**kwargs)
+        # if IPTS is set, also return its value
+        if self._ipts is not None:
+            ans["IPTS"] = self.IPTS
+        return ans
+
+    def json(self, **kwargs) -> str:
+        # get the dictionary and dump it
+        return json.dumps(self.dict(), **kwargs)
+
+    def __setattr__(self, name, value):
+        # set as usual
+        try:
+            super().__setattr__(name, value)
+        # if a key error occured, check if the key was IPTS
+        # if it was IPTS, then set the IPTS
+        # otherwise raise an error
+        except ValueError as e:
+            if name == "IPTS":
+                object.__setattr__(self, "_ipts", value)
+            else:
+                raise e
+
+    # if something calls for the IPTS, and it is not set,
+    # then call the GetIPTS mantid function to set the IPTS
+    @property
+    def IPTS(self):
+        from mantid.simpleapi import GetIPTS
+
+        if self._ipts is None:
+            self._ipts = GetIPTS(self.runNumber, Config["instrument.name"])
+        return self._ipts
```

Comment:
I really feel like these implementations stuffs shouldnt live in a dao.  I understand why its here but I think this really calls for a `GroceryService` to handle all the bits and bobs of IPTS.

Reply:
My goal was to make it completely transparent to the user that IPTS wasn't different from any other field in the grocery item, and also that it didn't need to be initialized until it was needed.  But... thinking on it more, and after doing all this work, it might not be needed at all.... I think FetchGroceriesRecipe can just call GetIPTS in the one place where the IPTS is needed...

Reply:
Though I suppose FetchGroceriesRecipe already interacts with disk directly so... Just keep in mind we may at some point need an alternate impl